pkg/gcp/pubsub.go

```diff
@@ -1,66 +1,59 @@
-	"bytes"
+	"cloud.google.com/go/pubsub"
 	"context"
-	"encoding/json"
-	"fmt"
-	"net/http"
-
-	"github.com/SAP/jenkins-library/pkg/log"
 	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"google.golang.org/api/option"
+	"os"
 )
 
-const api_url = "https://pubsub.googleapis.com/v1/projects/%s/topics/%s:publish"
-
-// https://pkg.go.dev/cloud.google.com/go/pubsub#Message
-type EventMessage struct {
-	Data        []byte `json:"data"`
-	OrderingKey string `json:"orderingKey"`
-}
-
-type Event struct {
-	Messages []EventMessage `json:"messages"`
+type PubsubClient struct {
+	projectID   string
+	pool        string
+	provider    string
+	topic       string
+	orderingKey string
 }
 
-func Publish(projectNumber string, topic string, token string, key string, data []byte) error {
-	ctx := context.Background()
-
-	// build event
-	event := Event{
-		Messages: []EventMessage{{
-			Data:        data,
-			OrderingKey: key,
-		}},
+func NewGcpPubsubClient(projectID, pool, provider, topic, orderingKey string) *PubsubClient {
+	return &PubsubClient{
+		projectID:   projectID,
+		pool:        pool,
+		provider:    provider,
+		topic:       topic,
+		orderingKey: orderingKey,
 	}
+}
 
-	// marshal event
-	eventBytes, err := json.Marshal(event)
+func (cl *PubsubClient) Publish(data []byte) error {
+	oidcToken := os.Getenv("PIPER_OIDCIdentityToken")
+	accessToken, err := GetFederatedToken(cl.projectID, cl.pool, cl.provider, oidcToken)
 	if err != nil {
-		return errors.Wrap(err, "failed to marshal event")
+		return errors.Wrap(err, "could not get federated token")
 	}
-	log.Entry().Debugf("created pubsub event: %s", string(eventBytes))
 
-	// create request
-	request, err := http.NewRequestWithContext(ctx, http.MethodPost, fmt.Sprintf(api_url, projectNumber, topic), bytes.NewReader(eventBytes))
-	if err != nil {
-		return errors.Wrap(err, "failed to create request")
-	}
+	return publish(cl.projectID, accessToken, cl.topic, cl.orderingKey, data)
+}
 
-	// add headers
-	request.Header.Set("Content-Type", "application/json")
-	request.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+func publish(projectNumber, accessToken, topic, orderingKey string, data []byte) error {
+	ctx := context.Background()
 
-	// send request
-	response, err := http.DefaultClient.Do(request)
+	staticTokenSource := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: accessToken})
+	pubsubClient, err := pubsub.NewClient(ctx, projectNumber, option.WithTokenSource(staticTokenSource))
 	if err != nil {
-		return errors.Wrap(err, "failed to send request")
-	}
-	if response.StatusCode != http.StatusOK {
-		return fmt.Errorf("invalid status code: %v", response.StatusCode)
+		return errors.Wrap(err, "pubsub client creation failed")
 	}
 
-	//TODO: read response & messageIds
+	msg := &pubsub.Message{Data: data, OrderingKey: orderingKey}
+	publishResult := pubsubClient.Topic(topic).Publish(ctx, msg)
+
+	// publishResult.Get() will make API call synchronous by awaiting messageId or error.
```

Comment:
We could make it asynchronous if called before Splunk. However, if Splunk is removed, this functionality will still become synchronous since it's the last call in the step and we have to await a response from GCP. I suggest we keep it synchronous for now, and if we identify any performance issues, we can consider refactoring.